### PR TITLE
History State API

### PR DESCRIPTION
**Pull request by cbrewster**
This PR implements the state-related portions of the History API. We need to think out how exactly to handle the history state in the case of 1. Reloads and 2. Pruning/Reloading Distant History.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15261)

### CODE REVIEW

File: tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/persisted-user-state-restoration/scroll-restoration-fragment-scrolling-samedoc.html.ini [hunk 1]
```diff
@@ -1,5 +1,6 @@
 [scroll-restoration-fragment-scrolling-samedoc.html]
   type: testharness
+  expected: TIMEOUT
   [Manual scroll restoration should take precedent over scrolling to fragment in cross doc navigation]
-    expected: FAIL
+    expected: TIMEOUT
```
Comment (cbrewster) [inline tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/persisted-user-state-restoration/scroll-restoration-fragment-scrolling-samedoc.html.ini]: This new `TIMEOUT` is due to `hashchange` events not being implemented yet.

File: tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/persisted-user-state-restoration/scroll-restoration-navigation-samedoc.html.ini [hunk 2]
```diff
@@ -1,5 +1,6 @@
 [scroll-restoration-navigation-samedoc.html]
   type: testharness
+  expected: TIMEOUT
   [history.{push,replace}State retain scroll restoration mode and navigation in the same document respects it]
-    expected: FAIL
+    expected: TIMEOUT
```
Comment (cbrewster) [inline tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/persisted-user-state-restoration/scroll-restoration-navigation-samedoc.html.ini]: Same as above

**Comment by cbrewster** [comment]
```
  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] History interface: attribute state

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] History interface: operation pushState(any,DOMString,DOMString)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] History interface: operation replaceState(any,DOMString,DOMString)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] History interface: window.history must inherit property "state" with the proper type (2)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] History interface: window.history must inherit property "pushState" with the proper type (6)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] History interface: calling pushState(any,DOMString,DOMString) on window.history with too few arguments must throw TypeError

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] History interface: window.history must inherit property "replaceState" with the proper type (7)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] History interface: calling replaceState(any,DOMString,DOMString) on window.history with too few arguments must throw TypeError
```

**Comment by cbrewster** [comment]
@jdm I have been looking at how to handle the `StructuredCloneData`. I have tried using an enum that stores either `StructuredCloneData` or a cached `Heap<JSVal>`. This enum is stored in a `HistoryEntry` which is stored in a `HashMap` on `History`. The issue is that every time I read from `StructuredCloneData`, it always gives me a garbage value. I noticed that Gecko is using a `StructuredCloneContainer`, I am not sure if that is just a wrapper over `StructuredCloneData` or if its doing something more complicated.

**Comment by cbrewster** [comment]
I determined that one way to deal with some of these intermittents was to bite the bullet and handle history entries for hash changes and firing the `hashchange` event. There are still some intermittent issues around the `history traversal task source`. I am unsure how to handle the task source since we are already having to queue a task to the constellation. Should we queue to the history traversal task source first and then have that queue a message to the constellation?

Also there are some tests failing due to improper structured clones... I don't believe it's anything I am doing in this PR, I think it can be investigated/fixed in a different PR.

Hopefully I can figure out a way to fix these last few intermittents so this can land.

**Comment by cbrewster** [comment]
The timing issues are do to the tests not waiting long enough for the history traversal to be processed by the constellation. We have a couple different options:
 1. Queue the traversal task on the event loop (using history traversal task source) and then sending a synchronous message to the constellation to do the traversal.
 2. Increase the timeout

any thought or other ideas? @asajeffrey

**Comment by jdm** [comment]
Script should never communicate synchronously with the constellation; this can lead to deadlock when the constellation communicates synchronously with the script thread.

**Comment by cbrewster** [comment]
@jdm true, I would like to avoid it at all costs. I wonder if it would be possible to rewrite these tests to utilize `popstate` rather than using a timeouts.

**Comment by asajeffrey** [comment]
@jdm I think it's okay for script to block waiting on the constellation, but we should never have the constellation blocked waiting on script. Indeed, we already have at least one case where script blocks waiting on the constellation: https://github.com/servo/servo/blob/master/components/script/dom/htmliframeelement.rs#L717. We may need to do something similar here.

**Comment by cbrewster** [comment]
@bors-servo try
Alphabet is hard

File: components/constellation/constellation.rs [hunk 3]
```diff
@@ -2125,7 +2186,8 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
             let evicted_id = frame.prev.len()
                 .checked_sub(PREFS.get("session-history.max-length").as_u64().unwrap_or(20) as usize)
                 .and_then(|index| frame.prev.get_mut(index))
-                .and_then(|entry| entry.pipeline_id.take());
+                .and_then(|entry| entry.pipeline_id);
+            let evicted_id = if Some(frame.pipeline_id) == evicted_id { None } else { evicted_id };
```
Comment (asajeffrey) [inline components/constellation/constellation.rs]: This will evict, for example, a pipeline where there are many entries after it with the same pipeline id. Should we not look to see if `frame.prev[N]` and `frame.prev[N+1]` have different pipeline ids?

File: components/constellation/constellation.rs [hunk 4]
```diff
@@ -2405,13 +2467,34 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
...
+                let result = match self.pipelines.get(&current_pipeline) {
+                    Some(pipeline) => {
+                        let msg = ConstellationControlMsg::RemoveStateEntries(current_pipeline,
+                            discarded_history_states);
+                        pipeline.event_loop.send(msg)
+                    },
+                    None => return warn!("Pipeline {} closed after removing state entries", current_pipeline),
+                };
+                if let Err(e) = result {
+                    self.handle_send_error(current_pipeline, e);
```
Comment (asajeffrey) [inline components/constellation/constellation.rs]: This code would be cleaner if we grouped the past by pipeline id, though it means we'd have to keep track of length separately.

**Review by asajeffrey — COMMENTED** [review]
First pass. TL;DR: I am confused about making traversal synchronous.

**Comment by asajeffrey** [comment]
Should the constellation be storing `StateId`s, or the serialized form of the state objects?

**Comment by asajeffrey** [comment]
If we do decide to use `StateId`s, can we just use indexes for them, since the constellation and the script thread agree about the number and order of state objects?

File: components/script/task_source/history_traversal.rs [hunk 5]
```diff
@@ -2,19 +2,67 @@
...
-    fn clone(&self) -> Box<ScriptChan + Send> {
-        box HistoryTraversalTaskSource((&self.0).clone())
+    fn handler(self: Box<HistoryTraversalRunnable>) {
+        let window = self.window.root();
+        let global_scope = window.upcast::<GlobalScope>();
+        let pipeline = global_scope.pipeline_id();
+        let (sender, recv) = ipc::channel().expect("Failed to create ipc channel");
+        let msg = ConstellationMsg::TraverseHistory(Some(pipeline), self.direction, sender);
+        let _ = global_scope.constellation_chan().send(msg);
+        let _ =  recv.recv();
```
Comment (cbrewster) [inline components/script/task_source/history_traversal.rs]: The idea here is that the tests use `setTimeout` to allow the event loop to have a chance to process the history traversal. This synchronous message makes sure the traversal goes through before the event loop has a chance to call the `setTimeout` callback. Although this is not a perfect solution, as we still have to wait for the async message back from the constellation to actually update the history state/url

File: components/constellation/constellation.rs (hunk 3 above)
Comment (cbrewster) [inline components/constellation/constellation.rs]: What I did instead was have `close_pipeline` handle setting entry pipeline ids to `None`.

**Comment by cbrewster** [comment]
The new plan of attack to handle the case of distant history being purged is to store serialized state data in the `constellation` and remove `StateID`s completely. We need to handle how to restore pipeline ids across entries that originally shared the same pipeline before they were purged. 

One possible option:
Entries always have a PipelineID, if the pipeline lookup fails, that means that the pipeline was purged and we then reload a new pipeline but reuse the same PipelineID. This would make all of the other entries that share the same PipelineId use the new pipeline rather than reloading once again.

Thoughts? @asajeffrey

**Comment by asajeffrey** [comment]
I'm not sure it should be the constellation doing this. Perhaps the resource manager, since it's already handling cookies?

Recycling the pipeline id might be the easiest thing to do, it means we'd be updating the `Document` of all entries with the same pipeline id. We'd need to check if there's a pending document for the id, as well as checking if it has an entry in the pipelines table.

Not sure what to do about a case where document A does a pushState to A', then navigates to B, A' is discarded, we go back to A' then before A' has matured, go back to A. We'd need to queue up the message to update the url and state object for A.

**Comment by cbrewster** [comment]
Reusing the same pipeline id does cause some other unintended issues as `Pipeline`s are not closed right away. They are closed after the constellation received the pipeline exited message from the script thread. Having an entry keeping the pipeline id causes some unintended messages to be send. There are also cases where we may discard a pipeline and traverse back to that entry that had that pipeline before the pipeline was completely closed.

We could create a new pipeline id and replace all entries that had the old pipeline id when we discard a pipeline. When we lookup this new pipeline id, no pipeline would be associated yet so we would trigger a new load.

**Comment by cbrewster** [comment]
Blocked on #15566

**Comment by asajeffrey** [comment]
Sounds like recycling pipeline IDs is having unexpected consequences.

**Comment by asajeffrey** [comment]
IRC conversation about session histories with multiple instances of the same pipeline id: http://logs.glob.uno/?c=mozilla%23servo&s=24+Feb+2017&e=24+Feb+2017#c619161

TL;DR: If you have document A which pushStates to B then again to C, then their shared Document is discarded, then the history is traversed to B, what happens? Our read of the spec is that A, B and C should all be reloaded to the same fresh document. This means that any session history entries with the same pipeline id must be contiguous.

(Duplicating https://github.com/servo/servo/issues/10992#issuecomment-282187366 just to make sure I don't forget.)

**Comment by nox** [comment]
What is the status on this? If this is stalled, could we close the PR and file an issue referencing it?

**Comment by cbrewster** [comment]
@nox right now we need to find a better way of handling distant history pipeline discard that allows the history to contain multiple entries with the same pipeline id. I'm finishing up school at the moment so I don't have time right now, but I plan on finishing this up here in a couple weeks.

I'll close for now until I have time to resume work.
